src/Aspire.Hosting/HttpService/HttpServiceBuilderExtensions.cs

```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Sockets;
+using System.Text.Json;
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting;
+
+/// <summary>
+/// Provides extension methods for adding Http Service resources to the application model.
+/// </summary>
+public static class HttpServiceBuilderExtensions
+{
+    /// <summary>
+    /// Adds a Http Service to the application model.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
+    /// <param name="uri"></param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{HttpServiceResource}"/>.</returns>
+    public static IResourceBuilder<HttpServiceResource> AddHttpService(this IDistributedApplicationBuilder builder, string name, string uri)
+    {
+        return builder.AddHttpService(name, new Uri(uri, UriKind.Absolute));
+    }
+
+    /// <summary>
+    /// Adds a Http Service to the application model.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
+    /// <param name="uri"></param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{HttpServiceResource}"/>.</returns>
+    public static IResourceBuilder<HttpServiceResource> AddHttpService(this IDistributedApplicationBuilder builder, string name, Uri uri)
+    {
+        var address = uri.GetLeftPart(UriPartial.Authority).Remove(0, uri.GetLeftPart(UriPartial.Scheme).Length);
+
+        var httpService = new HttpServiceResource(name);
+        return builder.AddResource(httpService)
+                      .WithAnnotation(new ManifestPublishingCallbackAnnotation(writer => WriteHttpServiceResourceToManifest(writer, uri)))
+                      .WithAnnotation(new AllocatedEndpointAnnotation(name, ProtocolType.Tcp, address, uri.Port, uri.Scheme));
```

Comment:
I think you can remove this.

Reply:
Right now I'm using `AllocatedEndpointAnnotation` and `ApplyBindings` to set the HttpService url in project environment variable in order for service discovery to use that.

https://github.com/dotnet/aspire/blob/57437728ad2aa53c5021d55f8aa372a265598000/src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs#L81

Should I use different or new Annotation. I think if we agree on what manifest.json should look like I will be able to complete this as expected.

Reply:
No don't use any annotation. Treat this resource like a custom type with a Uri property. Since we're adding an overload of WithReference that takes this concrete type, we can manually write the environment information instead of using this binding stuff. 

```C#
public class HttpServiceResource(string name, Uri uri) : Resource(name)
{
    public Uri Uri { get } = uri;
}

public class HttpResourceExtensions
{
    public static IResourceBuilder<HttpServiceResource> AddHttpService(this IDistributedApplicationBuilder builder, string name, Uri uri)
    {
        var httpService = new HttpServiceResource(name, uri);
        return builder.AddResource(httpService);
    }

public static IResourceBuilder<TDestination> WithReference<TDestination>(this IResourceBuilder<TDestination> builder, IResourceBuilder<HttpServiceResource> source)
    where TDestination : IResourceWithEnvironment
    {
        builder.WithEnvironment($"services__{source.Resource.Name}", source.Resource.Uri.ToString());
    }
}
```


Something like this.

Reply:
Note ... this will need the `_<number>` suffixes for those variable names to be picked up by service discovery I think.

Reply:
I think we support both